### ISSUE TITLE
Support OIDC identities for docker login

### ISSUE THREAD

**Issue by mlbiam** [docker/roadmap#314]
**Tell us about your request**
Currently dockerhub requires a static token or password.  These tokens are easily lost and hard to track and replace.  It would be much better for dockerhub to support an oidc identity that can be created by a pipeline as opposed to a static token.  These tokens can be validated by either a public `well-known` oidc discovery document or via a pinned certificate for private pipelines.

**Which service(s) is this request for?**
dockerhub

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I want to be able to securely publish containers from my GitHub actions pipeline with the pipeline's identity, rather then a static account stored as a secret.

**Are you currently working around the issue?**
Not yet, though planning to

**Additional context**

**Comment by martin-traverse** [comment]
Hi, any update on this? We're using OIDC for integrations with the cloud platforms, would be great to have that for pushing to Docker Hub as well.

Apologies if there is another way of doing this that I have missed!

**Comment by dotjoshrc** [comment]
Hey friends, I realize this is a very old ticket and we have been wanting to do this for quite a while now. Happy to say that we are starting planning and scoping out what an OIDC type of provider would look like for registry and hub. Nothing solid yet, but just an update 😄

**Comment by dotjoshrc** [comment]
Hello all! Just an update here. We have our first system design and are underway on this!

We hear ya on the "Bring my own auth for OIDC" and we'll get there, but the first implementation of this will be GitHub Actions OIDC to start. A lot of what we are designing is a more broad implementation so this can branch into whatever customers want to use.

Thanks for the patience and expect more updates in Q1 of this year. 🎉

**Comment by ajschmidt8** [comment]
@technicallyjosh, just checking in. Any updates here?

**Comment by scott-doyland-burrows** [comment]
It would be great to have OIDC available from GitHub Actions and Terraform Cloud, so we can terraform our Docker repos and teams etc, and not just push images.

**Comment by etugarinova** [comment]
Hi all! We're currently working on offering GitHub OIDC and will release it internally first. For BETA, were targeting somewhere mid July. Once we have more concrete dates, I'll post here.

**Comment by scott-doyland-burrows** [comment]
I'm assuming once it works for GitHub, then adding Terraform Cloud would be relatively simple and so could follow fairly soon after.

**Comment by JonZeolla** [comment]
@etugarinova if we want to participate in the beta, should we just subscribe to this issue or is there another way to submit?

**Comment by dotjoshrc** [comment]
> I'm assuming once it works for GitHub, then adding Terraform Cloud would be relatively simple and so could follow fairly soon after.

You are correct in your assumption. We plan long-term to integrate with other providers. Would be great to understand the most wanted for sure.

@JonZeolla We will be releasing as "Beta" to everyone.

I can update here: We are starting full internal testing here in the next week or so. Beta will start shortly after!

**Comment by mariusgrigaitis** [comment]
CircleCI please. They already support OIDC with AWS

https://circleci.com/docs/guides/permissions-authentication/openid-connect-tokens.html

**Comment by dotjoshrc** [comment]
> CircleCI please. They already support OIDC with AWS

Yup! Definitely on our list to check out :) We've built this to be pretty generic in authentication so any future OIDC providers should be way less work to set up for us.

Appreciate the call out.

**Comment by PhilippWendler** [comment]
Support for GitLab CI would be welcome. It already supports OIDC with PyPI.

**Comment by martin-traverse** [comment]
Sounds like this has been progressing - any ideas on the timeline for a public release yet?

NPM put their OIDC feature to GA a couple of months ago. We've had a renewed focus on moving everything onto OIDC following Shal Halud.

**Comment by zliang-akamai** [comment]
I'm wondering the available date of this feature too. Really appreciate it if we can get it earlier!

**Comment by bijoy26** [comment]
Would be a great use-case to support **Google Artifact Registry** as [Remote Repository / Pull-through Cache](<https://cloud.google.com/blog/products/devops-sre/using-authenticated-logins-for-docker-hub-in-google-cloud/>).

**Comment by acouvreur** [comment]
Very interested in this feature too, if there`s a beta test possible, my FOSS organization can participate: https://github.com/sablierapp

**Comment by kohtala** [comment]
> > CircleCI please. They already support OIDC with AWS
> 
> Yup! Definitely on our list to check out :) We've built this to be pretty generic in authentication so any future OIDC providers should be way less work to set up for us.
> 
> Appreciate the call out.

I'm curious. Is there anything against letting the user decide which JWT token `iss`, `aud`, `sub` to allow specified access to a repository?

**Comment by davidcorrigan714** [comment]
> Yup! Definitely on our list to check out :) We've built this to be pretty generic in authentication so any future OIDC providers should be way less work to set up for us.

You shouldn’t be setting up providers at all beyond a nice UI for configuring one or two of the most popular ones. We need to be able to configure any OIDC issuer. I know the current trend is for registries to support “trusted providers” but enterprises very often don’t use popular cloud providers for CI/CD. They need the capability to configure any OIDC issuer, like Hashicorp Vault, Azure Managed Identities, AWS IAM roles, or an issuer backed by an on-prem HSM. We happen to use Azure DevOps for CI/CD, which while a SaaS offering with OIDC tokens, registries haven’t supported it and get stuck still using legacy authentication methods and managing an inane number of tokens. Azure has a really good generic OIDC implementation in Workload Identity Federation that could be a good reference example. JFrog has a good OIDC implementation as well. AWS’s is workable ( like most of the IAM design it’s a bit wonky and doesn’t support standard endpoints 🤦‍♂️)

**Comment by alexaka1** [comment]
Following Sha1-Hulud v2 it would be nice to accelerate this a bit.

**Comment by zliang-akamai** [comment]
Hi there, any update?

**Comment by zliang-akamai** [comment]
Hi @dotjoshrc, is there any update on the internal testing? When will we get it for GitHub actions?

**Comment by dotjoshrc** [comment]
Hey all. So sorry for the delay of updates here. I can confirm that this work is being prioritized again. We were close to a GA late last year but some stuff had to move around. We are currently identifying the last bits that need to be addressed before going GA.

**Comment by yuneng-berri** [comment]
@dotjoshrc Thanks for your continued work on this!

+1 to the feature request. LiteLLM is very interested in moving away from long lived credentials.

**Comment by dotjoshrc** [comment]
@davidcorrigan714 I totally agree with you. We need to start somewhere. There is a ton of risk in just allowing "generic" OIDC and want to be careful with our posture around this considering the supply chain that relies on us. We'd like to tackle some specifics as an MVP and then 100% move towards the ability for you as the customer to say what is valid and what is not. I promise, this is not just some backlog item but rather more of a strategic approach that we are taking when it comes to "risk".

Sorry trying to catch up on comments here 😅

**Comment by alexaka1** [comment]
Obligatory "how many TeamPCP attacks until it's fine to ask for an update?" Oidc is not a be all end all solution, but it is incredibly needed. No pressure :)

**Comment by alexanderjohn** [comment]
Seeking update on this issue or a possible ETA for GA?

**Comment by juhoinkinen** [comment]
The [docker/login-action v4.5.0](https://github.com/docker/login-action/releases/tag/v4.5.0) introduced support for OIDC with PR https://github.com/docker/login-action/pull/1048.

**Comment by dotjoshrc** [comment]
Hey all, finally getting back around to this... It's finally released as mentioned above! 

https://docs.docker.com/enterprise/security/oidc-connections/

We really appreciate the patience on this one.

**Comment by alexaka1** [comment]
Thanks! Is it planned for individual accounts? Most supply chain attacks happen to single maintainer FOSS projects.

**Comment by nresare** [comment]
This is excellent progress. To make it usable for us, it would be lovely if we had the ability to trust not just tokens signed by the GitHub actions signing keys but to have the JWKS validation key endpoint be configurable. Would it make sense to track this feature request in this ticket, or should I open a new one?

**Comment by dotjoshrc** [comment]
> Thanks! Is it planned for individual accounts? Most supply chain attacks happen to single maintainer FOSS projects.

Anticipated but not planned yet. It's on our radar though as this has been a common feedback item so far.

> This is excellent progress. To make it usable for us, it would be lovely if we had the ability to trust not just tokens signed by the GitHub actions signing keys but to have the JWKS validation key endpoint be configurable. Would it make sense to track this feature request in this ticket, or should I open a new one?

There are ideas internally of how we'd accomplish this that are circulating. Nothing set in stone, but we definitely have this on our radar. Nothing concrete though on timelines or anything.

Of course, both feel free to open separate enhancement requests for this. We can track them on the IAM team as weighing requests.

**Comment by davidcorrigan714** [comment]
You should really update this roadmap title if you want other OIDC IdPs to be a separate feature, it's current "OIDC identities" which doesn't imply a narrowly scoped implementation specifically for GitHub Actions.

Practically there's tons of examples to implement this and not supporting generic OIDC federation is a pretty big gap for any sufficiently enterprisey platform ( The irony of GH itself not supporting OIDC not withstanding ). Almost all implementations are based on some basic globs on OIDC claims for authorization policy through a RFC 8693 based token exchange endpoint. Some practical examples:

Microsoft: https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation
JFrog: https://docs.jfrog.com/administration/docs/openid-connect-integration
AWS: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html
GCP: https://docs.cloud.google.com/iam/docs/workload-identity-federation

I do a lot of our platform integration work for IaC and developer operations - stuff that doesn't have OIDC really causes a ton of extra manual work for day-to-day work and a huge security risk and operation risk managing those credentials between rotation schedules, log leaks, overly broad permissions, & general token distribution management compared to OIDC service-to-service schemes. For example with Azure DevOps we'd be able to map the exact pipeline definition to which packages it can publish and that's super easy based on Azure DevOp's pipeline or service connection OIDC token. When the operational cost per token goes down we can afford to make much narrower tokens which inherently increases auditability of use and token provenance. We do a ton of OIDC auth for service-to-service for Azure, AWS, JFrog, Terraform & Azure DevOps and it's night-and-day in both operation effort and security posture.

**Comment by dotjoshrc** [comment]
@davidcorrigan714 you are absolutely right. I'm working with my product counterpart to separate custom/generic OIDC from GH OIDC. I think it started as a generic request, but the first step was GitHub as that's one of the most common requests and use-cases we have. More updates soon on this as it should be tracked separately.
